cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaSpecChecker.java

```diff
@@ -0,0 +1,154 @@
+     * their format version to match.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkKafkaLogMessageFormatVersion(List<Condition> warnings) {
+        String logMsgFormatVersion = kafkaCluster.getConfiguration().getConfigOption(KafkaConfiguration.LOG_MESSAGE_FORMAT_VERSION);
+        String kafkaBrokerVersion = spec.getKafka().getVersion();
+        if (logMsgFormatVersion != null && kafkaBrokerVersion != null) {
+            Matcher m = VERSION_REGEX.matcher(logMsgFormatVersion);
+            if (m.find() && !kafkaBrokerVersion.startsWith(m.group(0))) {
+                warnings.add(buildCondition("KafkaLogMessageFormatVersion",
+                                            "log.message.format.version does not match the Kafka cluster version, which suggests that an upgrade is incomplete."));
+            }
+        }
+    }
+
+    /**
+     * Checks for a single-broker Kafka cluster using ephemeral storage. This is potentially a problem as it
+     * means any restarts of the broker will result in data loss, as the single broker won't allow for any
+     * topic replicas.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkKafkaStorage(List<Condition> warnings) {
+        if (kafkaCluster.getReplicas() == 1 && usesEphemeral(kafkaCluster.getStorage())) {
+            warnings.add(buildCondition("KafkaStorage",
+                    "A Kafka cluster with a single replica and ephemeral storage will lose topic messages after any restart or rolling update."));
+        }
+    }
+
+    /**
+     * Checks for a single-node ZooKeeper cluster using ephemeral storage. This is potentially a problem as it
+     * means any restarts of the pod will cause the loss of cluster metadata.
+     *
+     * @param warnings List to add a warning to, if appropriate.
+     */
+    private void checkZooKeeperStorage(List<Condition> warnings) {
+        if (zkCluster.getReplicas() == 1 && usesEphemeral(zkCluster.getStorage())) {
+            warnings.add(buildCondition("ZooKeeperStorage",
+                    "A ZooKeeper cluster with a single replica and ephemeral storage will be in a defective state after any restart or rolling update. It is recommended that a minimum of three replicas are used."));
+        }
+    }
+
+    private boolean isEphemeral(Storage storage) {
+        return Storage.TYPE_EPHEMERAL.equals(storage.getType());
+    }
+
+    private boolean usesEphemeral(Storage storage) {
+        if (storage != null) {
+            if (isEphemeral(storage)) {
+                return true;
+            }
+            if (Storage.TYPE_JBOD.equals(storage.getType())) {
+                JbodStorage jbodStorage = (JbodStorage) storage;
+                return jbodStorage.getVolumes().stream().anyMatch(this::isEphemeral);
+            }
+        }
+        return false;
+    }
```

Comment:
I think these could be static, and probably `ModelUtil` is a better place for them (more likely to be reused there).